tests/frontend/test_program.py

```diff
@@ -402,6 +402,52 @@ class DummyCircuit(Compiler):
         with pytest.raises(program.CircuitError, match="program contains 3 modes, but the device 'None' only supports a 2-mode program"):
             new_prog = prog.compile(device=spec, compiler=DummyCircuit())
 
+    @pytest.mark.parametrize(
+        "measure_op", [
+            ops.MeasureFock(),  # MeasureFock
+            ops.MeasureHomodyne(phi=0),  # MeasureX
+            ops.MeasureHomodyne(phi=42),  # MeasureHomodyne
+            ops.MeasureHomodyne(phi=np.pi/2),  # MeasureP
+            ops.MeasureHeterodyne(),  # MeasureHD
+            ops.MeasureHeterodyne(select=0)  # MeasureHeterodyne
+        ],
+    )
+    def test_incorrect_number_of_measurements(self, measure_op):
+        """Test that an exception is raised if the compiler is called with a
+        device spec with an incorrect number of measurements"""
+
+        class DummyCircuit(Compiler):
+            """A circuit with 2 modes"""
+            interactive = True
+            primitives = {'MeasureHomodyne', 'MeasureHeterodyne', 'MeasureFock'}
+            decompositions = set()
+
+        # set maximum number of measurements to 2, and measure 3 in prog below
+        device_dict = {
+            "modes": {
+                "max": {
+                    "pnr": 2,
+                    "homodyne": 2,
+                    "heterodyne": 2
+                }
+            },
+            "layout": None, "gate_parameters": {}, "compiler": [None]
+        }
+        # set the target to simulon manually to avoid device spec being loaded from api;
+        # needed for the measurements check to run
```

Comment:
Hmm, isn't this achieved by having `connection=None`? It seems like other than that, `DeviceSpec.__init__` just stores the `target`.

Reply:
Not sure actually. I just noticed that the device spec was being updated when using `target="simulon"`, so the tests wouldn't work, so I added this as a solution.

Reply:
That's odd 🤔 Maybe there's something else going one there? As done here, it should be a matter of setting an attribute 🤔

Reply:
It works as you said. I was probably just testing it locally, forgetting the `compiler=None`, when it didn't work (and didn't look into it). Removed it now. 🙂